packages/accounts-controller/CHANGELOG.md

```diff
@@ -7,6 +7,18 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## [17.2.0]
+
+### Added
+
+- Add `accountAdded` and `accountRemoved` events ([#4496](https://github.com/MetaMask/core/pull/4496))
+- Export `AccountsControllerListMultichainAccounts`,`AccountsControllerGetSelectedMultichainAccount`,`AccountsControllerGetNextAvailableAccountName}Action` actions ([#4497](https://github.com/MetaMask/core/pull/4497))
+
+### Fixed
+
+- Use `listMultichainAccounts` for non-EVM specific methods ([#4494](https://github.com/MetaMask/core/pull/4494))
+- Set `lastSelected` for initial account ([#4494](https://github.com/MetaMask/core/pull/4494))
```

Comment:
```suggestion
## [18.0.0]

### Added

- **BREAKING:** `AccountsControllerMessenger` must allow the new internal events `AccountsController:accountAdded` and `AccountsController:accountRemoved`. ([#4496](https://github.com/MetaMask/core/pull/4496))
  - Add and export corresponding event types `AccountsControllerAccountAddedEvent`, `AccountsControllerAccountRemovedEvent`. 
- Export action types `AccountsControllerListMultichainAccounts`,`AccountsControllerGetSelectedMultichainAccount`,`AccountsControllerGetNextAvailableAccountName}Action`. ([#4497](https://github.com/MetaMask/core/pull/4497))

### Fixed

- Use `listMultichainAccounts` instead of `listAccounts` for non-EVM specific multichain methods. ([#4494](https://github.com/MetaMask/core/pull/4494))
- For initial account, emit `selectedAccountChange ` and update `lastSelected`. ([#4494](https://github.com/MetaMask/core/pull/4494))
```

Reply:
I think the Fixed entries could be more user-friendly, and less reliant on knowledge of internal variables, but I'm lacking the understanding to come up with better phrasing. @montelaidev could you maybe provide some context on the user-facing changes here?

Reply:
@mcmire Should we see the addition of new internal actions/events as a breaking change? My initial reasoning was that the added actions/events now need to be present in the generic arguments of the non-restricted controller-messenger, but I'm not sure about this.

> Previous discussion: https://github.com/MetaMask/core/pull/4413#discussion_r1636655127

Reply:
@MajorLift That's the way I've thought about it too: the reason it's a breaking change is that consumers now needs to allowlist that action or event when initializing the messenger, otherwise an error will be thrown when attempting to call or subscribe to that action or event. The fact that the messenger doesn't throw immediately if a used action or event is omitted from the allowlist is unfortunate, but there will be an error eventually.

Reply:
For the fixed entries, they are only internal changes that fixes a bug.

Reply:
@MajorLift Oh, I just realized you're talking about internal actions/events. Consumers don't need to explicitly allowlist internal actions/events, right? So then this would not be breaking after all. Is that right?

Reply:
For restricted messengers, only allowlists need to be supplied, but for unrestricted controller-messengers, consumers still define them by enumerating all actions/events e.g.

```ts
new ControllerMessenger<
  | AccountsControllerActions 
  | NetworkControllerActions 
  | NetworkControllerAllowedActions, 
  ...,
>();
```

The part I'm confused about is whether to call this non-breaking because users have access to the union types that fully enumerate internal actions/events, or to call this breaking because users could choose to use individual action/event type exports, in which case they'd need to keep track of additions to the internal actions/events e.g.

```ts
new ControllerMessenger<
//   | TokensControllerGetStateAction
  | TokensControllerAddDetectedTokensAction
  | TokensControllerExampleNewAction,
  ...,
>();
```

Reply:
Hmm, good point. I am inclined in this case not to treat it as a breaking change. My reasoning is that even if a consumer left out `AccountsControllerAccountAddedEvent` and `AccountsControllerAccountRemovedEvent` from the type parameter for the unrestricted controller messenger, the controller itself would still be able to emit the event. Whether the consumer is actually subscribing to the event or not is on the consumer, but I don't believe they'd be required to do that in order to upgrade. Does that make sense?

Reply:
Hmm I was going to argue that adding these new internal actions/events will cause type errors for the `messenger` constructor option in downstream instantiation calls of the accounts controller (more type properties -> more requirements -> narrower type object: `A & B <: A` i.e. contravariant). 

However, I've found that restricted messengers with incomplete or even empty internal actions/events lists are being accepted by our controller constructors, as long as they have compatible allowlists for external actions/events. I'll have to create a ticket for addressing this.

I think it's accurate to say, at least for now, that this wouldn't be a breaking change for consumers. Once the controller has been instantiated, I agree that it makes sense to only look at whether the controller can publish the new events, and have the controller be agnostic about whether any given messenger (other than its own messagingSystem) subscribes to those events.

Reply:
I put together a [playground repro](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBDAnmApnA3nAUHOBBAYwIgFcA7GAZwGEIKoIAbRlKQmYOygGh3yNIUadGA2asAogDcUQ3rkLFyVWvSYsoAcRQxFg3QQ51u2BQOXC14rToDKKFoZQATPcvacyJvm6GrR6qz2jjAu0iC+MNQAFgCGZADmKNKyMLwAvnAAZgwgcADkAAIgOrEgsZQA1gD0seZCALTEVhr5WKCQsBim-Ep+ImIaGdm5BcWl5VW19VRNA4FQbR3Q8EioGHwAKhCVspYB1towtjCxoR7GWzt7-oNBp6Ex8UkpFPJw27tkAEpnKPt3NiGTw8HqfWS-UIAhavKjeXDgn5-aGHOwPFAXLxg65IqG3BYnP5PRLJGRvLCZHIQPJFEqnSY1CqUHSUOYtViUJbgFbdRGQ-746xwSmjWkTCqMyjM2bNA4aTntbldTCCjQAWX+zJJUGFI2pYzpZQl1QARhUUGy5awuZ1Vsg0Og+ABpFCIKDARKq1hHF1uj0JABi0EimPevvdnvmqJg4f9lAAQohNvbQ87XRGEl6bAY+lRU7hY5H2VBCY84iTYcMqTTxvTjbs-YlLYCbTy1g64AA5HQAd2glSzR1LGOBxi7vf7g503ZgfagA8YwFSiYAkq5R1iZ3OB1GNMPiS8yfARfqxXWpmQJ-PmwtW11291bGRYmB9+WkrrqwbxVNKM+wKysqApyADcWBYNUABUkE4JBcAAIwAHRwFmcDNJQogkIY0BwKwDCXqQlCMIgcB1AQKBgFQcAlFKshJDqwBZHAdBoDhIDQKxTHAFRHqhFAz6MNRmqxB+i4YZQpFQGgABEKDgEg0mwdU4FQfgyHxhAMDRHAvGsAJQlSiJaBiVRsRSbh8kkZBylYOh8ByZRiArhQemxIwGqGUkAAywDiVmHlavRcAALxwJePYoburABXRrAADx8NU1RwAABnyyJTsc6KYilOkSVprFkMRQkgCarDMUxaU4vyKIaJilApXwEgAB4EIwJDOCgcXpXiUVAkYZCgj1Ap9UO2UbgAfHAAA+HzVRlfV4MwEA9i49XvElqXDbV9xEu+pKpLlvkINEhXFSUpXlRAlXbVmsINU1rXtZ13Xzb1xb3SYt19W+zwHRQU2zd9xZLYwK1hEelBYBNAAUACUfBYOFc1fDVWYw46uA0YFrAAFwWY5zl8QJAVGT5fl9TF2qIUkMDfP8ojAE4zgY3wuDPiU+P5MDVqLO8uBuWDq3rgNlD4wA2mzuAFIiO1QLjtPDvk-PS-kW6Tn1CvTleC5LhQq7OMrUu4GrOtZlrWV-Eb0uq5EcsW5E1s2wUdvm4rDgoMzjsqwAuirgvg8490S8bMs4vbGF7X9Ts26bs4a8WuOR2W0cqybrua8yIQQxEMwHigbQ237fDpHD7wwDiAAK7rkZQ9hQFITMoPjJokMAjDOGqEAEJUiLV03desI35HwxkCMQXB8bIQA8kVJEsTpLn8W5BmUEZcAmXlBNIHA1ngXZ29OUvJPCd5vlRJTmqxTqoXI-5V-aglCJvSNIMbhJQMv3LoOB+tiXJVVVGC1iy-QrEeI6+VTrMTniVMqOprpbS-ndSGjVcAtTah1LqPNASfRRhCYBvNQGHlSIDPBuJX68x-sLe60N4aI2RtghYrMsYP3ovjBySAiauXcqfFA5ML7FipvRGmOh6YYRrqEFmmNpYc2bmHIB71eYx1IstYW9UQ7O3kTcTWit0TKJNurecbttbxyMYuZcTlDZpwKIYnciddFW2sfkDO9idDe1Ds4mY9t3bZ3XLmZRxcbYB2oZDDRztubh3NsnFA+d9E2LNpndEsSnEuN5knD2zNwiRFiVLQJwoy58Arl8futd67Dzka3dundu69yrjXf4ZSm6jwpOPTaeAoAJBICUCgFUED2gKNJRh1gLbDmkvkLeZBNKkSlMABIz4TQsAQBAOAYAzJlB0FdJiD4ILJXyIMyJOi0R-GkjNOA0lbHGJgBc8x+tLEnNmuchJrjLahHudgTa0lUmAgdjMN5nyvGXOCJ7SRkQxmIRhgAJgAMwABYACsCMOFH2Jm5UmZ8KaCNYawRCBBBYwwiQoih3yHGhHyIiyyXDl48M8nw8+cshHYtxcwfFFzDkvILq05K7TOndPgAgh8eyhkaCTkk-aYyJlTKZLM+ZiyK4rLWXSTZfT1g7IGbLKJoq-p-NZYnaJ+c-lfIWOk3xWS85ivyOC6F8LyWE2Pqi3h-D760WppQEgJpKAEHdGVfFQq8Z6v2srOA4sfbjwnjBXecAIXIVQnZKAWEK46igLEXy-xcJQAYDqHsp0yDytos4Ve184A9m4tpaSHpiDgBYK8xeKLBLY3XiZKGu8bIH3LdSMAVaUCUpPjSx1l9nVBVvigCKTqcZQCfqclqog6gwAnTbQ11h1o9BtrNBdGgjghg3Mu6Wq6AWjTsBkkFMxMRSymnwXduZv6qLWu-d4s0p1JsMHO6Wa7JCQ2sa+ksh6c7ZP2rCU9PQL36CvULCGqQoaw3HsjT9rNNoIn6fkUd184qfo3cerdQGLCZSBV7dDA13nJUw-0EG16RYghMJ+nDkjTW5nzrCU5n6qFgaEBNcZqrjqTPgFKuZsQFloDlQKz9DLFiIVVbgauEBUCwBIvkAAxGAd0UgraL2VWgRD-ax3jKklkDkSzSJwGcIxHTUkekXTgQRk6Zw0LxE43AMqpEiBX3zV+YtWkPSqYKGIxmzMkPagtZCqFEKIUIzaXu4siFnCISoDDeCAAGCFJgACccN8abCgSgZqqBmYeYramr8CnJOsB3vkbG182ObSzUzbSx1OptTMi4OAp1zJ0A8-kTApXtT4yE1iqAIE4DRK62Fwh6I+vpELgWzri8K0do2d2+1vbz6+fou8AbGwba6SpXbfG0igleO2+kaxWdgU3tzFzZRB2S5jxUlyjpXTUi9IFf8y9gL0Tio45KmZPG+N6dWUmxV8Ctn9NVXsz9PzcwGqG8Sg9vjQWnOkoWTMbKEdBiBODizDyEeXIRwmJM9o-mY7ZXbMFAXrXtDINNztc3qVr3RQI3mwmcV4s8c9tlSsEZtsrbNu11OyaLY09fRnzLmfAcud7AoUIcz6DXGS8CYaDMUVkJ1cnS4JJ7ywA+X1bBSP1RCoBshcsxp-ExHrnVvMjjXL1jAA2J7cCzVN4CQ35wNznt6CL-dkv3DO9t67rD7uqMnf0JiMC6v+ma6Y0HSGuvcAu-VT9TVYDUgu-twSePxCKAu8o9+iPudaN-qPMHh8jHtfvxCi7gnxYfTpn9CjzdA0y9V6LGbnQ2PEzJlQDb055em8e6EEH8ChfIcLHD-dXXLunwviISgevjZEcgNT-9GAwesjkC3ZUjuXce593qYPBuTS4ZrbgFJGAJB+IH4FpQRA5Psg6AINELfA94Zn+lkfk-Ob0DpDAjbC7NslOLmcFHD0NcWwEgMAW0FwGGAAfVv2TTIDXHxnIEqEmR7DIBSzgCgLiEAPzWOhSli2agABJ0BxF-R0hcodtn8dBX8EA40UBP9pZv9pZf9gB-9HgT8TMCBEBgDQCVhwCoDWDZB2D4CyBECVoUD8ZeD01+CSJjoiDEgn9cAX9T9MIaCpZ6D0gKRwIgA) for this. 

Looks like constructor controllers do raise an error if they're passed a messenger that was defined with an incomplete list of internal actions/events, but fail to raise that error if either the `Action` or `Event` type union don't contain any of the controller's internal actions/events. Interesting...